SGD minimizer

yorkerlin opened the pull request:
@karlnapf 
take a look at this minimizer (plain sgd)

yorkerlin commented:
ref
https://github.com/shogun-toolbox/shogun/issues/2779
https://github.com/shogun-toolbox/shogun/issues/2853

yorkerlin commented:
@karlnapf 
the design is change to include `L2penlty` and `momentum`.

I will soon update this PR once unit tests are passed.

yorkerlin commented:
the design enable users to specify `learning rate`, `penalty`, and `momentum update`.

yorkerlin commented:
@karlnapf 
I will soon send two PRs about
* SGD with L2penalty  and SGD with momentum update 
* svrg minimizer (ref http://riejohnson.com/svrg_download.html)

karlnapf commented:
Hey, this is too much to review, could you split?

yorkerlin commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.h:
```diff
@@ -0,0 +1,120 @@
+ *  decorator pattern
+ *
+ */
+class CFirstOrderPenalizedCostFunction: public CFirstOrderCostFunction
```
@karlnapf 
take a look at this.
It also related to the `cost function design`

yorkerlin commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.h:
```diff
@@ -0,0 +1,120 @@
+	 *
+	 * @param penalty_type the type of penalty
+	 */
+	virtual void set_penalty_type(CPenalty* penalty_type);
```
the different penalty types (L2 and L1) can be plugged in

karlnapf replied:
Should this really be another class? CPenalty?

yorkerlin commented on src/shogun/optimization/FirstOrderPenalizedStochasticCostFunction.h:
```diff
@@ -0,0 +1,130 @@
+{
+/** @brief The first order penalized stochastic cost function base class.
+ *
+ * note that the cost function must be the finite sum of sample-specific cost
```
this is class for  sample-specific cost functions
the type of cost functions can be written as
f(w)= \sum_i{f_i(w)}
where i is the index of samples

For example,
lease squares loss
f(w)=0.5 * \sum_i{ (y_i-w'*x_i)^2 },
where (x_i,y_i) are features and a label  for the i-th sample.

karlnapf replied:
I understand

yorkerlin commented on src/shogun/optimization/FirstOrderCostFunction.h:
```diff
@@ -0,0 +1,68 @@
+	 *
+	 * @return gradient of variables
+	 */
+	virtual CMap<TParameter*, SGVector<float64_t> >* get_gradient()=0;
```
@karlnapf 
One approach:
* if we use `SGVector<float64_t>` instead of `CMap<TParameter*, SGVector<float64_t> >*`, all issues in https://github.com/shogun-toolbox/shogun/pull/2872 are gone.
* We can provide some static helper functions (eg in `CMath.h`) to do conversion between  `SGVector<float64_t>` and `CMap<TParameter*, SGVector<float64_t> >*`,

Another approach
* keep `CMap<TParameter*, SGVector<float64_t> >*`
* We do need help functions to do  conversion between  `SGVector<float64_t>` and `CMap<TParameter*, SGVector<float64_t> >*`
* We also need a piecewise weighted `add`  functions  to do `add(CMap<TParameter*, SGVector<float64_t> >* a, float a_weight, CMap<TParameter*, SGVector<float64_t> >*b, float64_t b_weight)`, which means `a_weight*a+b_weight*b`

Let me know your suggestion.

karlnapf replied:
Yeah this CMap thing seems a bit of an overkill. Are there any computational implications?

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	SG_UNREF(opt);
+}
+
+TEST(SGDMinimizer,test4)
```
@karlnapf 
this is an example about how to use this framework

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	aa->set_x(x);
+	aa->set_y(y);
+	aa->set_init_w(w);
+	CMyStochasticCostFunction *fun=new CMyStochasticCostFunction();
```
step1 define/implement one stochastic cost function (that is a finite sum of sample-specific cost function)
eg:
f(w)=0.5* \sum_i{ (y_i-w'*x_i)^2 }, where (x_i,y_i) are features and a label for the i-th sample, w is the objective variable.

lisitsyn replied:
It would make sense to put that as a comment into the code

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	CMyStochasticCostFunction *fun=new CMyStochasticCostFunction();
+	fun->set_target(aa);
+
+	CFirstOrderPenalizedStochasticCostFunction* fun2=new CFirstOrderPenalizedStochasticCostFunction(fun);
```
step2 add a penalized cost function (decorator pattern)

lisitsyn replied:
Same here, better push it as a comment

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+
+	CFirstOrderPenalizedStochasticCostFunction* fun2=new CFirstOrderPenalizedStochasticCostFunction(fun);
+	fun2->set_penalty_weight(1.0);
+	CL2Penalty* penalty_type=new CL2Penalty();
```
L2 penalty

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	CFirstOrderPenalizedStochasticCostFunction* fun2=new CFirstOrderPenalizedStochasticCostFunction(fun);
+	fun2->set_penalty_weight(1.0);
+	CL2Penalty* penalty_type=new CL2Penalty();
+	fun2->set_penalty_type(penalty_type);
```
register penalty

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+
+	CSGDMinimizer* opt=new CSGDMinimizer(fun2);
+
+	CConstLearningRate* rate=new CConstLearningRate();
```
learning rate

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	CSGDMinimizer* opt=new CSGDMinimizer(fun2);
+
+	CConstLearningRate* rate=new CConstLearningRate();
+	rate->set_const_learning_rate(0.0001);
```
register learning rate

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+
+	CConstLearningRate* rate=new CConstLearningRate();
+	rate->set_const_learning_rate(0.0001);
+	CGradientDescendWithMomentumUpdater* updater=new CGradientDescendWithMomentumUpdater(rate);
```
gradient updater

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	rate->set_const_learning_rate(0.0001);
+	CGradientDescendWithMomentumUpdater* updater=new CGradientDescendWithMomentumUpdater(rate);
+	updater->set_momentum(0.9);
+	opt->set_gradient_updater(updater);
```
step5 register gradient updater

yorkerlin replied:
note thtat
for batch update, step5 and step6 can be viewed as linear search methods
for online update, step5 and step6 should be separated

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	CL2Penalty* penalty_type=new CL2Penalty();
+	fun2->set_penalty_type(penalty_type);
+
+	CSGDMinimizer* opt=new CSGDMinimizer(fun2);
```
step 4 which minimizer to be used

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.cc:
```diff
@@ -0,0 +1,458 @@
+	float64_t cost=opt->minimization();
+	cost=(cost-aa->get_cost())+aa->get_cost()/y.vlen;
+
+	//the result is from the svrg software using plain stochastic gradient descent
```
reference result

yorkerlin commented on tests/unit/optimization/SGDMinimizer_unittest.h:
```diff
@@ -0,0 +1,83 @@
+};
+
+
+class CRegressionExample: public CSGObject
```
this is a problem we try to solve.
since we do not do multiply inheritance, we need to define its cost function separately.

yorkerlin commented on src/shogun/optimization/SGDMinimizer.cpp:
```diff
@@ -0,0 +1,80 @@
+	SG_REF(variable_reference);
+	CFirstOrderStochasticCostFunction *fun=dynamic_cast<CFirstOrderStochasticCostFunction *>(m_fun);
+	REQUIRE(fun,"the cost function must be a stochastic cost function\n");
+	for(;m_cur_passes<m_num_passes;m_cur_passes++)
```
we can cache some results if samples are read from a hard drive

yorkerlin commented on src/shogun/optimization/SGDMinimizer.cpp:
```diff
@@ -0,0 +1,80 @@
+	for(;m_cur_passes<m_num_passes;m_cur_passes++)
+	{
+		fun->begin_sample();
+		while(fun->next_sample())
```
this step can be paralleled (either synchronous or asynchronous)
for now it just a simple implementation.

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,100 @@
+
+namespace shogun
+{
+/** @brief This is a learning rate class for a descent-base minimizer.
```
descent-based

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,100 @@
+	/*  destructor */
+	virtual ~CConstLearningRate() {}
+
+	/** returns the name of the class
```
As usual: Capital letters: "Returns the name " ....
Makes the doxygen look much nicer

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,100 @@
+
+	/** set the const learning rate
+	 *
+	 * @param learning_rate learning_rate must be positive and usually is not greater than 1.0
```
Usually?
Details?
Shall we not enforce it?

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,100 @@
+	 */
+	virtual void set_const_learning_rate(float64_t learning_rate)
+	{
+		REQUIRE(learning_rate>0.0, "learning_rate (%f) must be positive",learning_rate);
```
"Learning rate" capital L

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,100 @@
+	 */
+	virtual float64_t get_learning_rate()
+	{
+		REQUIRE(m_const_learning_rate>0.0,"learning_rate must set\n");
```
This error message does not make sense

karlnapf commented on src/shogun/optimization/LearningRate.h:
```diff
@@ -0,0 +1,58 @@
+ * This is an interface class used in descent based minimizers.
+ *
+ */
+class CLearningRate: public CSGObject
```
Why would there be a seperate class for learning rates?
Just creates tons of overhead: C++ code of the base class, etc etc, this seems to be just a float.

lisitsyn replied:
yeah, +1

yorkerlin replied:
@lisitsyn 
for first order gradient based minimizer,
we have 
* penalty (L1, L2 and elasticnet)
* learning rate (const, many line search methods, inverse scaling)
* gradient update (gradient descend, momentum update,  rprop)

if we want combination of all three things, class should be a good choice.
we can use static function pointer to do so.

yorkerlin replied:
@karlnapf 
@lisitsyn 

for learning rate, function pointer is not a good choice too.
Reasons:

*for `inverse scaling learn rate`, `learning_rate = eta0 / pow(t, power_t)`, where `t` is the times we call get_learning rate, `eta0` and `power_t` must be given by users or default value.

for class, we can store `t`, `eta0` and `power_t` as data members.

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,64 @@
+	 * @param variable_reference a reference of the target variable
+	 * @param negative_descend_direction the negative descend direction given the current variable
+	 */
+	virtual void update_variable(CMap<TParameter*, SGVector<float64_t> >* variable_reference,
```
Classes with just one virtual methods should not inherit from CSGObject in my eyes. See my comments in the other PR. 

@lisitsyn can you suggest a better way here? You know all the good patterns.

karlnapf replied:
See also the LearningRate.h class

lisitsyn replied:
Classes with one method are functions ;) Not sure what's best in this case

yorkerlin replied:
for gradient updater, sadly, function pointer is not a good choice.

yorkerlin replied:
for learning rate, function pointer is not a good choice too.
Reasons:
* invscaling learn rate `eta = eta0 / pow(t, power_t)`, where t is the times we call `get_learning rate`. 

for class, we can store it as data member.

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,109 @@
+	 * 
+	 * @return optimal value
+	 */
+	virtual float64_t minimization()=0;
```
minimize()

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,109 @@
+	 * 
+	 * @return whether minimizer supports batch update
+	 */
+	virtual bool support_batch_update() const=0;
```
supports_batch_update

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,109 @@
+	{
+		if(m_fun!=fun)
+		{
+			SG_UNREF(m_fun);
```
first REF then UNREF

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,109 @@
+	 *
+	 * @param fun the cost function
+	 */
+	virtual void set_cost_function(CFirstOrderCostFunction *fun)
```
This method is not require to be visible from SWIG, right?

karlnapf commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,172 @@
+
+void CFirstOrderPenalizedCostFunction::set_cost_function(CFirstOrderCostFunction* fun)
+{
+	REQUIRE(fun,"cost function must not be NULL\n");
```
Capital C

karlnapf commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,172 @@
+		"number (%d) of objective variables must match number (%d) of their gradients\n",
+		m_variable_reference->get_num_elements(),grad->get_num_elements());
+
+	for(index_t i=0; i<m_variable_reference->get_num_elements(); i++)
```
auto

karlnapf commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,172 @@
+	for(index_t i=0; i<m_variable_reference->get_num_elements(); i++)
+	{
+		CMapNode<TParameter*, SGVector<float64_t> >* var=m_variable_reference->get_node_ptr(i);
+		REQUIRE(var->key->m_name,"the name of parameter not set\n");
```
The

karlnapf commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,172 @@
+		for(; j<grad->get_num_elements(); j++)
+		{
+			CMapNode<TParameter*, SGVector<float64_t> >* g=grad->get_node_ptr(j);
+			REQUIRE(g->key->m_name, "The name of parameter not set \n");
```
whitespace

karlnapf commented:
Hey @yorkerlin before I go on looking at this.

We need to change the workflow a bit. It is impossible to review this code. Too many files, too many classes. I also don't really get why we need this massive amount of new objects here. I see that you want to keep things flexible, but the code is quite cryptic -- not so much the individual files, the the collection of classes all together.

I suggest
1. Class diagram that describes the high-level structure (no details yet)
2. Discussion, also with other devs. Cleaning up the design. @lisitsyn and @lambday should also have a look.
3. Updates and iterating over 1-2

4. Seperate PRs for the base classes, with documentation focussing on the high-level interaction between classes.
5. Atomic PRs for the individual implementations, along with unit tests

6. Benchmarking and notebooks

We cannot continue to merge more of this stuff in the current way -- it has gotten too complicated. I realise you want to push on implementing the cool algorithms, but I ask you to slow down a bit and make sure the fundamental framework is good before filling it with algorithms.

lisitsyn commented:
@karlnapf @yorkerlin can we discuss it a bit more at #shogun?

lisitsyn commented:
Some thoughts:

1. I feel it would be better to untie this thing from shogun structures. It would lead to cleaner API without maps and stuff. C-style API is much cleaner sometimes.
2. I lack understanding of use cases of this kind of library? It seems reasonable for me to describe the code we want to get first. 
3. Its very OOP now while some functions could suffice. One of the best approaches is non-member functions operating on simple objects.

yorkerlin commented:
@lisitsyn 
@karlnapf 
this framework will offer optimizers for gp inference using strategy pattern.

once we defined the cost function,  this framework will can be extend to do the following things.
* http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html  
* https://github.com/JuliaOpt/Optim.jl

yorkerlin commented:
@karlnapf 
@lisitsyn 

ref:
http://yaroslavvb.blogspot.ca/2014/03/stochastic-gradient-methods-2014.html
my goal is to implement 
* svrg minimizer (done)
* `Accelerated, Parallel and Proximal Coordinate Descent` for Lasso (L1 penalty)

yorkerlin commented:
@lisitsyn 
non-member functions 
do you mean `function pointer`?

yorkerlin commented:
@karlnapf 
@lisitsyn 
take a look at this

yorkerlin commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,126 @@
+
+void CFirstOrderPenalizedCostFunction::set_cost_function(CFirstOrderCostFunction* fun)
+{
+	//REQUIRE(fun,"cost function must not be NULL\n");
```
@karlnapf 
how can I use `REQUIRE` if the class is not a subclass of `SGObject`

yorkerlin commented on src/shogun/optimization/FirstOrderCostFunction.h:
```diff
@@ -0,0 +1,65 @@
+ * This class gives the interface used in a first-order gradient-based unconstrained minimizer
+ *
+ */
+class CFirstOrderCostFunction
```
@karlnapf 
Now, the `cost function` is `not` a subclass of `SGObject`.

karlnapf replied:
yep, as we discussed. Thats good now, we can add as many of those classes as we like

yorkerlin commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,102 @@
+		m_fun=fun;
+	}
+
+	virtual CMinimizerContext* save_to_context()=0;
```
@karlnapf 
@lisitsyn 
we return a `CMinimizerContext` (which is a subclass of SGObject) to store the mutable variables.

karlnapf replied:
I see. Thats actually a good way to do this!

yorkerlin commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,102 @@
+
+	virtual CMinimizerContext* save_to_context()=0;
+
+	virtual void load_from_context(CMinimizerContext* context)=0;
```
we load from a `CMinimizerContext` (which is a subclass of SGObject) to get the mutable variables.  (deserialization)

yorkerlin commented on src/shogun/optimization/FirstOrderCostFunction.h:
```diff
@@ -0,0 +1,65 @@
+	 *
+	 * @return reference of variables
+	 */
+	virtual SGVector<float64_t> obtain_variable_reference()=0;
```
@karlnapf 
now, we use SGVector instread of CMap.
Once we agree on this framework, I will update the PRs https://github.com/shogun-toolbox/shogun/pull/2874 and https://github.com/shogun-toolbox/shogun/pull/2872

karlnapf replied:
great!

This "variable reference" thing is what exactly? Should be documented, I find it confusing

yorkerlin replied:
For target variable (for minimization), we need the `in place` modification for these variables.

For the corresponding gradient, we do not need the `in place` requirement.

yorkerlin commented on src/shogun/optimization/FirstOrderPenalizedCostFunction.cpp:
```diff
@@ -0,0 +1,126 @@
+	//REQUIRE(m_variable_reference.vlen>0,"the reference of objective variable must set\n");
+	float64_t cost=m_fun->get_cost();
+
+	for(auto idx=0; idx<m_variable_reference.vlen; idx++)
```
@karlnapf 
travis complains the `auto` keyword.
ref
https://travis-ci.org/shogun-toolbox/shogun/jobs/73320014

yorkerlin commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+		float64_t penalty=0.0;
+		if(m_penalty_type)
+		{
+			for(auto idx=0; idx<var.vlen; idx++)
```
@karlnapf 
travis complains about the `auto` keyword.
ref
https://travis-ci.org/shogun-toolbox/shogun/jobs/73320014

In fact, we can use the `linalg` here.

karlnapf replied:
@vigsterkr @lisitsyn the auto keyword breaks travis. What to do?

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,66 @@
+{
+/** @brief This is a base class for descend update.
+ *
+ * the base class do the following job:
```
"The" base class "does" the following job:

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,66 @@
+ * the update_variable function will do
+ * variable_reference-=update_variable(negative_descend_direction)
+ *
+ * Note that an example of a negative_descend_direction is gradient  
```
"is to simply use the gradient, or XXX"

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,66 @@
+class CDescendUpdater
+{
+public:
+	/**update the variable_reference based on the negative_descend_direction
```
I think doxygen cannot parse this if there is no space after the *  -- but not sure.
In any case. "Update" with capital U

@yorkerlin I know this is minor and it must be annoying that I constantly moan about this, but let's just use proper English for the interfaces ;)

yorkerlin replied:
@karlnapf 
I know. 
Currently, the goal of this PR is about class design.
I will update the document.

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,66 @@
+	 * @param variable_reference a reference of the target variable
+	 * @param negative_descend_direction the negative descend direction given the current variable
+	 */
+	virtual void update_variable(SGVector<float64_t> variable_reference,
```
why is this called "variable_reference"?

yorkerlin replied:
`in place` modification

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,66 @@
+
+	virtual void update_context(CMinimizerContext* context)=0;
+
+	virtual void load_from_context(CMinimizerContext* context)=0;
```
undocumented?
What would this doo?

yorkerlin replied:
currently, I focus on class design.
Comments will be update soon.

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+{
+
+/** @brief The first order minimizer base class.
+ *
```
Please elaborate a bit more here. I don't understand what this class does without reading code. The class description should at least give an idea of what it is used for

karlnapf replied:
I like the new description, very helpful.

Again it would be good to put the class in context and reference other classes via doxygen links

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+	 */
+	virtual void set_cost_function(CFirstOrderCostFunction *fun)
+	{
+		m_fun=fun;
```
No reference counting here, but a NULL check maybe?

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+	virtual void load_from_context(CMinimizerContext* context)=0;
+
+
+	/** set the weight of penalty
```
"Set"

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+	 */
+	virtual void set_penalty_weight(float64_t penalty_weight)
+	{
+		//REQUIRE(penalty_weight>0,"penalty_weight must be positive\n");
```
what about the REQUIRE?

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+ * This class gives the interface of a minimizer
+ *
+ */
+class CFirstOrderMinimizer
```
If the class does not inherit from CSGObject, please remove the "C" prefix.

@lisitsyn do you agree?

lisitsyn replied:
yes

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+
+	/** Destructor */
+	virtual ~CFirstOrderMinimizer()
+	{}
```
Should the m_fun not be deleted? Or something? How do you plan do manage such references?

yorkerlin replied:
For now, users/developers have to delete these points outside.

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+	{
+		if(m_penalty_type)
+		{
+			for(auto idx=0; idx<var.vlen; idx++)
```
parallelise this loop maybe?

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,161 @@
+		m_penalty_weight=0;
+	}
+};
+
```
OK for this class. Just the issue of memory managment remains (the not-Shogun classes)

yorkerlin replied:
users/developers have to delete all pointers.

karlnapf replied:
See my other comment and #2880

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,53 @@
+#include <shogun/optimization/FirstOrderStochasticCostFunction.h>
+namespace shogun
+{
+/** @brief The first order stochastic average cost function base class.
```
Doing what exactly?

karlnapf commented on src/shogun/optimization/FirstOrderStochasticCostFunction.h:
```diff
@@ -0,0 +1,64 @@
+{
+/** @brief The first order stochastic cost function base class.
+ *
+ * The class gives the interface used in first order stochastic minimizers
```
Please elaborate what this does from a high-level perspective, the example is good!

karlnapf replied:
Tell people what a first order stochastic cost function is (they might not know)

karlnapf commented on src/shogun/optimization/FirstOrderStochasticCostFunction.h:
```diff
@@ -0,0 +1,64 @@
+	 */
+	virtual void begin_sample()=0;
+
+	/* get next sample
```
"Get" ...

karlnapf commented on src/shogun/optimization/FirstOrderStochasticMinimizer.h:
```diff
@@ -0,0 +1,148 @@
+namespace shogun
+{
+
+/** @brief The base class for stochastic first-order gradient-based minimizers.
```
Document!

karlnapf commented on src/shogun/optimization/FirstOrderStochasticMinimizer.h:
```diff
@@ -0,0 +1,148 @@
+	 */
+	virtual void set_gradient_updater(CDescendUpdater* gradient_updater)
+	{
+		//REQUIRE(gradient_updater, "gradient_updater must set\n");
```
?

karlnapf commented on src/shogun/optimization/FirstOrderStochasticMinimizer.h:
```diff
@@ -0,0 +1,148 @@
+	{
+		//REQUIRE(gradient_updater, "gradient_updater must set\n");
+
+		m_gradient_updater=gradient_updater;
```
Again the question of reference counting.

karlnapf replied:
See #2880

karlnapf commented:
I generally agree with most of this. One thing I just thought is that a way to reduce overhead when adding Shogun classes would be to have a hierarchy of base classes with different functionality. See #2880

@yorkerlin One thing that needs to be worked out is how to manage memory with these new classes. Somehow needs to delete instances .... who would that be?

Apart from that, there are a few minors but these base classes look good to merge (after minors are resolved)

yorkerlin commented:
@karlnapf 
take a look at the update comments.

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,105 @@
+{
+/** @brief This is a learning rate class for a descent-based minimizer.
+ *
+ * This class gives a const learning rate during descent update.
```
Please mention the other classes that would use this class and how, using doxygen references

karlnapf replied:
Also pls explain the context mechanism here, referencing the methods

yorkerlin replied:
see the updated comments

karlnapf commented on src/shogun/optimization/ConstLearningRate.h:
```diff
@@ -0,0 +1,105 @@
+ *
+ */
+
+#define IGNORE_IN_CLASSLIST
```
very good idea!

karlnapf commented on src/shogun/optimization/DescendUpdater.h:
```diff
@@ -0,0 +1,77 @@
+#include <shogun/optimization/MinimizerContext.h>
+namespace shogun
+{
+/** @brief This is a base class for descend update.
```
Good!

But also here: pls reference all the classes that play together with this one

yorkerlin replied:
see updated comments

karlnapf commented on src/shogun/optimization/FirstOrderCostFunction.h:
```diff
@@ -0,0 +1,80 @@
+ * This class gives the interface used in a first-order gradient-based unconstrained minimizer
+ *
+ * For example: least square cost function \f$f(w)\f$
+ * \f[
```
References to related classes, explain the interface, referencing the obtain_variable_context method

yorkerlin replied:
see updated comments

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,206 @@
+	}
+
+	/** Constructor
+	 * @param fun cost function (user have to manully delete the pointer)
```
What do you think about having a Shogun base class without serialisation and all the stuff. ONLY reference counting.
Then all your classes could be subclasses of that one and users would not have to do things.

See #2880

karlnapf replied:
How you did it here is fine, but it is a source of memory leaks

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,206 @@
+	 */
+	virtual float64_t minimize()=0;
+
+	/** Does minimizer support batch update
```
question mark at the end

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,206 @@
+	 */
+	virtual void set_penalty_type(Penalty* penalty_type)
+	{
+		if(m_penalty_type!=penalty_type)
```
why the if?

karlnapf commented on src/shogun/optimization/FirstOrderMinimizer.h:
```diff
@@ -0,0 +1,206 @@
+	 * \f[
+	 * L2f(w)=f(w) + L2(w) \f]
+	 * where \f$ f(w)=\sum_i{(y_i-w^T x_i)^2}\f$ is the least sqaure cost function
+	 * and \f$L2(w)=\lambda \frac{w^t w}{2}\f$ is the L2 penalty
```
good!

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,102 @@
+/** @brief The class is about a stochastic cost function for stochastic average minimizers.
+ *
+ * The cost function must be Written as a finite sample-specific sum of cost.  
+ * For example, least squares cost function,
```
Class references pls, otherwise good

karlnapf commented on src/shogun/optimization/FirstOrderSAGCostFunction.h:
```diff
@@ -0,0 +1,102 @@
+namespace shogun
+{
+/** @brief The class is about a stochastic cost function for stochastic average minimizers.
+ *
```
Pls explain the interface, referencing the main methods of this class

yorkerlin replied:
see updated comments

karlnapf commented on src/shogun/optimization/FirstOrderStochasticCostFunction.h:
```diff
@@ -0,0 +1,91 @@
+	/** Get the SAMPLE gradient value wrt target variables 
+	 *
+	 * WARNING
+	 * This method does return 
```
good !

karlnapf commented on src/shogun/optimization/FirstOrderStochasticMinimizer.h:
```diff
@@ -0,0 +1,191 @@
+ * FirstOrderStochasticMinimizer uses a sample gradient \f$\frac{\partial f_i(w) }{\partial w}\f$ to find optimal target variables,
+ * where the index \f$i\f$ is generated by some distribution.
+ *
+ * Note that FirstOrderMinimizer uses the exact gradient \f$\frac{\partial f(w) }{\partial w}\f$. 
```
Very nice description.
Could you make sure doxygen produces links to the other classes you mention (google the syntax to reference other classes)

karlnapf commented on src/shogun/optimization/GradientDescendUpdater.cpp:
```diff
@@ -0,0 +1,99 @@
+{
+	REQUIRE(m_learning_rate,"learning_rate must set\n");
+	REQUIRE(context, "context must set\n");
+	m_learning_rate->load_from_context(context);
```
I have the feeling we should discuss this issue once more on IRC.

karlnapf commented on src/shogun/optimization/L2Penalty.h:
```diff
@@ -0,0 +1,82 @@
+#include <shogun/lib/config.h>
+namespace shogun
+{
+/** @brief The class is about L2 penalty/regularization.
```
This class implements L2 regularisation within the CLASS_REFERENCE framework

karlnapf commented on src/shogun/optimization/LearningRate.h:
```diff
@@ -0,0 +1,69 @@
+namespace shogun
+{
+/** @brief The base class about learning rate for descent-based minimizers.
+ *
```
Again, pls reference other classes that use this one and explain the high level interface

yorkerlin replied:
see updated commnets

karlnapf commented on src/shogun/optimization/Penalty.h:
```diff
@@ -0,0 +1,74 @@
+namespace shogun
+{
+/** @brief The base class for penalty/regularization used in minimization.
+ *
```
Elaborate, recferences

yorkerlin replied:
see updated commnets

karlnapf commented:
Hi @yorkerlin 

Very nice, we are getting there.
While I agree with you that this PR is about class design, the interface documentation is very important in that -- because it helps me understand what is going on, etc. I always see that as a part of the class design.

I am quite happy with the current set up. A few things need discussion

 * memory management of the classes
 * The serialisation seems slightly cumbersome with these chains of the "store context" calls

I have the feeling that we can maybe resolve both issues via putting in more structure in the Shogun base classes. For example, reference counting can definitely be solved. See the issue (feel free to participate in discussions there).

I think this is almost mergable though, but we have to keep the above points in mind and should resolve them eventually.

Very nice framework otherwise. Really like that.

karlnapf commented:
@yorkerlin It turns out CSGObject inherits from SGRefObject, which you can use as a base class for all your new classes. Then they have automatic reference counting (SG_REF/SG_UNREF), but nothing else.

This solves the above issues

karlnapf commented:
https://github.com/shogun-toolbox/shogun/blob/9305aa9d1bab28d20979033b0fd7c8e2d545b85e/src/shogun/base/SGObject.h

yorkerlin commented:
@karlnapf 
the old `CSGObject` inherits from `SGRefObject`.
However, the current `CSGObject` does `not` inherit from `SGRefObject`.

karlnapf commented:
Yeah I know, just pointing out that we can solve things this way.
Just requires some minimal work. But you can go withou the reference counting for now and then we update. Good?

yorkerlin commented:
@karlnapf 
Traivs still complains about the `auto` keyword.
https://travis-ci.org/shogun-toolbox/shogun/jobs/75011240

yorkerlin commented:
@karlnapf 
This framework can be extended to parallel/asynchronous computing and dealing with sparse vectors.

karlnapf commented:
@lisitsyn you added the auto. Any ideas?
@vigsterkr might also know

karlnapf commented:
@yorkerlin good point with the sparse vectors

karlnapf commented:
This is getting there, ready to merge in my eyes. 
If the guys dont reply about the auto problem, lets just change it back to standard and merge

yorkerlin commented on src/shogun/optimization/GradientDescendWithNAGUpdater.h:
```diff
@@ -0,0 +1,97 @@
+#include <shogun/optimization/GradientDescendWithMomentumUpdater.h>
+namespace shogun
+{
+/** @brief The class implements the gradient descend with the Nesterov Momentum (Nesterov's Accelerated Gradient) update.
```
@karlnapf 
I added a new gradient updater. 
Please take a look

yorkerlin commented:
@karlnapf 
since `plain gradient descend`, `rmsprop` , `adadelta`, and `adam` can use the `Momentum` corrections, I refactored the `updater` class.

The existing classes are designed based on the `generic` and `flexible` principle.
Since these classes may be not very efficient, I added a extra layer in the class  hierarchy  so that user can use efficient implementation if needed.

yorkerlin commented:
@karlnapf 
@iglesias 
errors reported by travis are not related to this PR.
I merge this PR
